### ISSUE TITLE
Getting it working with --disable-proto

### ISSUE THREAD

**Issue by MatthiasKunnen** [gustavnikolaj/hijackresponse#19]
Node 12 added the [--disable-proto](https://nodejs.org/docs/latest-v12.x/api/cli.html#cli_disable_proto_mode) flag to combat prototype pollution. This library, however, uses `__proto__`. I have tried to fix this with [Object.setPrototypeOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) but have not found a solution yet.

**Comment by gustavnikolaj** [comment]
Hi @MatthiasKunnen 

Thanks for the PR #18  and for your suggestion.

The nature of this module is quite on the hack-ish side. It works really well, and has been battle-tested running for years in a web application with 50k concurrent users at peak hours, but I almost want to say that the use of `--disable-proto` and using anything written based on hijackresponse is fundamentally incompatible. 😅 

What hijackresponse does is basically to monkeypatch the response object it is passed, such that data written to it no longer flows to the socket, but to hijackresponse itself instead. I can try to change the implementation to not use assignments to `__proto__`, but it will still be violating the flag - at least in spirit :-)

I haven't been looking at this module for a while, but since I am still using it in production for an app that I just upgraded that to node.js v14, I should probably take some time to brush this module off :-)

**Comment by MatthiasKunnen** [comment]
>  The nature of this module is quite on the hack-ish side

I've noticed when looking at the code 😄

What the flag does is basically removing things like `obj.__proto__ = foo` and `obj.__proto__.bar = foo`. Using methods such as [Object.setPrototypeOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) and [Object.create](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create) is still allowed. The purpose of the flag is not to remove `__proto__` altogether but prevent accidental vulnerabilities so you don't need to worry about violating the flag in spirit :)

That being said, no pressure about implementing this, it's a nice to have. I'm actually trying to get some changes merged in an upstream package after which I will no longer be required to hijack.

**Comment by gustavnikolaj** [comment]
Thanks for the additional context, I will definitely look into supporting the flag. Also nice to know that it's not urgent for you to have a fix :-) I'll keep the issue on the books for when I have better time to polish the project up.

The way hijackresponse is typically used (at least to my knowledge) is in post-processing middleware.

A good example is [express-compiless](https://www.npmjs.com/package/express-compiless) which will highjack responses for urls matching `/\.less$/` and run them through the less-compiler, and changing the content-type for `text/css`. This allows you to reference a less file in your html through a regular `<link>`-tag. That is really nice (if you have a build system that is also able to take the html file, discover the referenced less file and compile it) as it allows you to have a development workflow without a build-step.

There's also middlewares like [express-processimage](https://www.npmjs.com/package/express-processimage) which has more of a production-focus: It allows you to modify an image response, by using query parameters; e.g. `https://example.com/image.jpg?format=png&resize=100,100&crop&gravity=center` - that will result in a png response with a 100 by 100 pixel image cropped from the original response. That is useful when you have user supplied content and need to generate on the fly responsive images, or plain old boring thumbnails :-) With hijackresponse you can do it, without having to reimplement the `serve-static` module from express core.

If you use koa instead of express, you don't need hijackresponse at all, but can just do this instead:

```js
app.use((ctx, next) => {
  await next();

  // now you have access to req and res on the ctx object, after middleware lower in the
  // stack have had their hands on it - effectively equivalent to what hijackresponse gives
  // you for express
}
```

I am interested in hearing more about how you used it, as it seems that your use-case is different than what I've typically seen hijackresponse used for. By the sounds of it, you have used it to compensate for a bug in another middleware?

**Comment by MatthiasKunnen** [comment]
Those are some cool uses :+1: 

My use case is [express-graphql](https://github.com/graphql/express-graphql) which sets the status code of the response and finalizes the request without any way to customize it. <https://github.com/graphql/express-graphql/issues/427> has more information.

FYI I have created types for `hijackresponse` here: <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47540>

**Comment by gustavnikolaj** [comment]
This should now be addressed by #24 and the new release 5.0.0-rc.1 (under the next dist tag for now). There's an API change in the major release too.
